core/blockchain.go

```diff
@@ -937,11 +948,114 @@ func (bc *BlockChain) WriteBlockWithoutState(block *types.Block, td *big.Int) (e
+// caching it if found.
+func (bc *BlockChain) GetAccumulatedEpochUptime(epoch uint64) []istanbul.Uptime {
+	uptime := rawdb.ReadAccumulatedEpochUptime(bc.db, epoch)
+	if uptime == nil {
+		return nil
+	}
+	return uptime
+}
+
+// WriteAccumulatedEpochUptime stores the accumulated uptime for an epoch into the database, also caching it
+// along the way.
+func (bc *BlockChain) WriteAccumulatedEpochUptime(epoch uint64, uptime []istanbul.Uptime) error {
+	rawdb.WriteAccumulatedEpochUptime(bc.db, epoch, uptime)
+	return nil
+}
+
+// https://stackoverflow.com/questions/19105791/is-there-a-big-bitcount/32702348#32702348
+func bitCount(n *big.Int) int {
+	count := 0
+	for _, v := range n.Bits() {
+		count += popcount(uint64(v))
+	}
+	return count
+}
+
+// Straight and simple C to Go translation from https://en.wikipedia.org/wiki/Hamming_weight
+func popcount(x uint64) int {
+	const (
+		m1  = 0x5555555555555555 //binary: 0101...
+		m2  = 0x3333333333333333 //binary: 00110011..
+		m4  = 0x0f0f0f0f0f0f0f0f //binary:  4 zeros,  4 ones ...
+		h01 = 0x0101010101010101 //the sum of 256 to the power of 0,1,2,3...
+	)
+	x -= (x >> 1) & m1             //put count of each 2 bits into those 2 bits
+	x = (x & m2) + ((x >> 2) & m2) //put count of each 4 bits into those 4 bits
+	x = (x + (x >> 4)) & m4        //put count of each 8 bits into those 8 bits
+	return int((x * h01) >> 56)    //returns left 8 bits of x + (x<<8) + (x<<16) + (x<<24) + ...
+}
+
+func updateUptime(uptime []istanbul.Uptime, blockNumber uint64, bitmap *big.Int, window uint64, epochNum uint64, epochSize uint64) []istanbul.Uptime {
+	var validatorsSize uint64
+	if len(uptime) == 0 {
+		// the number of validators is upper bound by 3/2 of the number of 1s in the bitmap
+		// TODO: due to this maybe we create an uptimeScore array which is 1 element bigger than required
+		// is this a problem?
+		validatorsSize = uint64(math.Ceil(float64(bitCount(bitmap)) * 1.5))
+		// TODO: there is no need for more than the specific block if we know when the validator signed last itme
+		uptime = make([]istanbul.Uptime, validatorsSize)
+	}
+
+	valScoreTallyFirstBlockNum := istanbul.GetValScoreTallyFirstBlockNumber(epochNum, epochSize, window)
+	valScoreTallyLastBlockNum := istanbul.GetValScoreTallyLastBlockNumber(epochNum, epochSize)
+
+	// signedBlockWindowLastBlockNum is just the previous block
+	signedBlockWindowLastBlockNum := blockNumber - 1
+	signedBlockWindowFirstBlockNum := signedBlockWindowLastBlockNum - (window - 1)
+
+	for i := 0; i < len(uptime); i++ {
+		if bitmap.Bit(i) == 1 {
+			// update their latest signed block
+			uptime[i].LastSignedBlock = blockNumber - 1
+		}
+
+		// If we are within the validator uptime tally window, then update the validator's score if it's last signed block is within
+		// the lookback window
+		if valScoreTallyFirstBlockNum <= blockNumber && blockNumber <= valScoreTallyLastBlockNum {
```

Comment:
LGTM, nit: we could _maybe_ extract this and the `valScoreTally{First, Last}BlockNum` functions together to another function, `isInLookbackWindow`